pkg/server/server.go

```diff
@@ -0,0 +1,255 @@
+package server
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"log/slog"
+	"net"
+	"os"
+	"runtime"
+	"sync"
+	"time"
+
+	jobv1 "github.com/kralicky/jobserver/pkg/apis/job/v1"
+	"github.com/kralicky/jobserver/pkg/auth"
+	"github.com/kralicky/jobserver/pkg/jobs"
+	"github.com/kralicky/jobserver/pkg/rbac"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/keepalive"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+type Options struct {
+	ListenAddress   string
+	CaCertFile      string
+	CertFile        string
+	KeyFile         string
+	AuthMiddlewares []auth.Middleware
+}
+
+type jobInfo struct {
+	jobs.Process
+	owner  auth.AuthenticatedUser
+	cancel context.CancelCauseFunc
+}
+
+type Server struct {
+	Options
+	jobv1.UnsafeJobServer
+	jobs    sync.Map // map[string]jobInfo
```

Comment:
It would reduce lock contention since entries are only ever added, and never modified or deleted. Individual jobs are unrelated, so we don't necessarily need to block other clients from looking up existing jobs at the same time as a new one is being created.

In the pathological case, an argument might also be made that this could help keep request latency from being affected by things happening outside of what a given user is authorized to know.

Reply:
`sync.Map` also being good for concurrent, append-only maps is neat, I wasn't aware of that. 👍